### PR TITLE
Git - better handle symbolic links

### PR DESCRIPTION
**Pull request by lszomoru**
If we detect that the workspace folder contains a symbolic link, and the user is running `git 2.31` or later, we use the `--path-format` [flag](https://git-scm.com/docs/git-rev-parse#Documentation/git-rev-parse.txt---path-formatabsoluterelative) to get the path of the repository root path relative to the workspace folder. This enables us to set the repository root to a path that matches the workspace folder.

Related #5970

### CODE REVIEW

File: extensions/git/src/git.ts [hunk 1]
```diff
@@ -520,6 +520,13 @@ export class Git {
 			}
 		}
 
+		// Handle symbolic links
+		// Git 2.31 added --path-format flag to rev-parse which allows us to get the relative path of the repository root
+		if (!pathEquals(repositoryPath, repoPath) && !isDescendant(repoPath, repositoryPath) && !isDescendant(repositoryPath, repoPath) && this.compareGitVersionTo('2.31.0') !== -1) {
```
Comment (joaomoreno) [inline extensions/git/src/git.ts:525]: In order to make this more readable, should we rename `repositoryPath` to `pathInsidePossibleRepository` and `repoPath` to `repositoryRootPath`?

File: extensions/git/src/model.ts [hunk 2]
```diff
@@ -931,12 +931,14 @@ export class Model implements IBranchProtectionProviderRegistry, IRemoteSourcePu
...
 
-		const result = await Promise.all(workspaceFolders.map(async folder => {
-			const workspaceFolderRealPath = await this.getWorkspaceFolderRealPath(folder);
-			return workspaceFolderRealPath ? pathEquals(workspaceFolderRealPath, repositoryPath) || isDescendant(workspaceFolderRealPath, repositoryPath) : undefined;
-		}));
+		// The repository path may be a canonical path or it may contain a symbolic link so we have
+		// to match it against the workspace folders and the canonical paths of the workspace folders
+		const workspaceFolderPaths = new Set<string | undefined>([
+			...workspaceFolders.map(folder => folder.uri.fsPath),
+			...await Promise.all(workspaceFolders.map(async folder => await this.getWorkspaceFolderRealPath(folder)))
```
Comment (joaomoreno) [inline extensions/git/src/model.ts:938]:
```suggestion
			...await Promise.all(workspaceFolders.map(folder => this.getWorkspaceFolderRealPath(folder)))
```

**Review by joaomoreno — DISMISSED** [review]
Interesting!

> that the workspace folder contains a symbolic link

Can you scaffold an example scenario for this? Just some good setup steps in bash so we can recreate the setup we want to accurately test here.

**Comment by lszomoru** [comment]
Here is the macOS setup that I have used to validate these changes:
1. Open Terminal and navigate to your desktop
2. mkdir repro
3. cd repro && mkdir foo
4. cd foo && mkdir bar
5. cd bar && git init
6. code-insiders .
7. Add a new file (ex: README.md) with some content
8. Commit the changes
9. Close VS Code
10. Create the symbolic link in the Terminal
    * ln -s /Users/<USERNAME>/Desktop/repro/foo/bar /Users/<USERNAME>/Desktop/repro/bar
11. Navigate to the repro folder in the Termial
12. code-insiders bar
13. Open README.md and change some content
  * Notice that the dirty diff markers do not appear
14. Switch to the Source Control view and click on README.md
  * Notice how a second copy of the same file is opened

Issues listed in 13) and 14) are fixed with the changes from this PR.

**Comment by lszomoru** [comment]
@joaomoreno, thanks for the simplified steps and thanks for the feedback. As discussed privately, I will explore another approach where the model will continue to use absolute canonical paths to make sure that we do not open multiple git repositories pointing to the same location. Will close this PR for now and will reopen if needed.

**Comment by lszomoru** [comment]
After considering other approaches, we decided to pursue the approach implemented in this pull request.

**Review by joaomoreno — APPROVED** [review]

**Pull request merged**